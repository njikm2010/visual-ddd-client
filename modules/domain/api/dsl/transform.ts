/**
 * DSL 转换
 */
export {};
