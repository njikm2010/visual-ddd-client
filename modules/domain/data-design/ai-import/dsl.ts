import { DataObjectReferenceCardinality, DataObjectTypeName } from '../dsl/dsl';

export namespace AIImport {
  export interface DataObjectReferenceType {
    type: DataObjectTypeName.Reference;
    target: string;
    property: string;
    cardinality: DataObjectReferenceCardinality;
  }

  export interface DataObjectNormalType {
    type: Exclude<DataObjectTypeName, DataObjectTypeName.Reference>;
  }

  export interface DataObjectProperty {
    name: string;
    title: string;
    primaryKey?: boolean;
    type: DataObjectNormalType | DataObjectReferenceType;
  }

  export interface DataObject {
    name: string;
    title: string;
    properties: DataObjectProperty[];
  }
}
