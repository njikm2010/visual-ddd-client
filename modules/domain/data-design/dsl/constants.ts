import { DataObjectIndexMethod, DataObjectIndexType, DataObjectReferenceCardinality, DataObjectTypeName } from './dsl';

export enum DataObjectName {
  DataObject = 'dataObject',
}

export const DataObjectReadableName: Record<DataObjectName, string> = {
  [DataObjectName.DataObject]: '数据对象',
};

export const DataObjectTypeList: { name: DataObjectTypeName; label: string }[] = [
  { name: DataObjectTypeName.Boolean, label: '布尔型' },
  { name: DataObjectTypeName.Integer, label: '整型' },
  { name: DataObjectTypeName.Decimal, label: '十进制' },
  { name: DataObjectTypeName.Long, label: '长整型' },
  { name: DataObjectTypeName.Double, label: '双精度' },
  { name: DataObjectTypeName.Float, label: '浮点型' },
  { name: DataObjectTypeName.Date, label: '日期型' },
  { name: DataObjectTypeName.DateTime, label: '日期时间型' },
  { name: DataObjectTypeName.Timestamp, label: '时间戳型' },
  { name: DataObjectTypeName.String, label: '字符串' },
  { name: DataObjectTypeName.Text, label: '文本型' },
  { name: DataObjectTypeName.LongText, label: '长文本型' },
  { name: DataObjectTypeName.JSON, label: 'JSON型' },
  { name: DataObjectTypeName.Reference, label: '引用型' },
];

export const DATE_FORMAT = 'YYYY-MM-DD';
export const DATE_TIME_FORMAT = 'YYYY-MM-DD HH:mm:ss';

export const DataObjectIndexTypeName: Record<DataObjectIndexType, string> = {
  [DataObjectIndexType.Normal]: '普通索引',
  [DataObjectIndexType.Unique]: '唯一索引',
  [DataObjectIndexType.FullText]: '全文索引',
};

export const DataObjectIndexListList: { name: DataObjectIndexType; label: string }[] = [
  { name: DataObjectIndexType.Normal, label: DataObjectIndexTypeName[DataObjectIndexType.Normal] },
  { name: DataObjectIndexType.Unique, label: DataObjectIndexTypeName[DataObjectIndexType.Unique] },
  { name: DataObjectIndexType.FullText, label: DataObjectIndexTypeName[DataObjectIndexType.FullText] },
];

export const DataObjectIndexMethodList: { name: DataObjectIndexMethod; label: string }[] = [
  { name: DataObjectIndexMethod.BTREE, label: 'BTREE' },
  { name: DataObjectIndexMethod.HASH, label: 'HASH' },
];

/**
 * 支持主键的字段类型
 * @param type
 * @returns
 */
export function objectTypeThatSupportPrimaryKey(type: DataObjectTypeName) {
  return type !== DataObjectTypeName.JSON;
}

/**
 * 支持默认值的字段类型
 * @param type
 * @returns
 */
export function objectTypeThatSupportDefaultValue(type: DataObjectTypeName) {
  return (
    type === DataObjectTypeName.Boolean ||
    type === DataObjectTypeName.Integer ||
    type === DataObjectTypeName.Decimal ||
    type === DataObjectTypeName.Long ||
    type === DataObjectTypeName.Double ||
    type === DataObjectTypeName.Float ||
    type === DataObjectTypeName.Date ||
    type === DataObjectTypeName.DateTime ||
    type === DataObjectTypeName.Timestamp ||
    type === DataObjectTypeName.String
  );
}

/**
 * 反方向基数
 */
export const DataObjectReferenceCardinalityReversed: Record<
  DataObjectReferenceCardinality,
  DataObjectReferenceCardinality
> = {
  [DataObjectReferenceCardinality.ManyToMany]: DataObjectReferenceCardinality.ManyToMany,
  [DataObjectReferenceCardinality.OneToOne]: DataObjectReferenceCardinality.OneToOne,
  [DataObjectReferenceCardinality.OneToMany]: DataObjectReferenceCardinality.ManyToOne,
  [DataObjectReferenceCardinality.ManyToOne]: DataObjectReferenceCardinality.OneToMany,
};

/**
 * 可被引用的字段类型
 * @param type
 * @returns
 */
export function objectTypeThatReferable(type: DataObjectTypeName) {
  return type !== DataObjectTypeName.Reference && type !== DataObjectTypeName.JSON;
}

/**
 * 支持递增的字段类型
 * @param type
 * @returns
 */
export function objectTypeThatSupportAutoIncrement(type: DataObjectTypeName) {
  return (
    type === DataObjectTypeName.Integer ||
    type === DataObjectTypeName.Long ||
    type === DataObjectTypeName.Double ||
    type === DataObjectTypeName.Float
  );
}

export function objectTypeThatSupportLength(type: DataObjectTypeName) {
  return type === DataObjectTypeName.String;
}
