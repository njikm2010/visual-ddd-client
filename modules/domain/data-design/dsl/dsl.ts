import { NameDSL } from '@/modules/domain/domain-design/dsl/dsl';

/**
 * 类型
 */
export enum DataObjectTypeName {
  Boolean = 'Boolean',
  Date = 'Date',
  DateTime = 'DateTime',
  Timestamp = 'Timestamp',
  Integer = 'Integer',
  Decimal = 'Decimal',
  Long = 'Long',
  Double = 'Double',
  Float = 'Float',
  String = 'String',
  Text = 'Text',
  LongText = 'LongText',
  JSON = 'JSON',
  Reference = 'Reference',
}

/**
 * 布尔值
 */
export type DataObjectBoolean = {
  type: DataObjectTypeName.Boolean;
  defaultValue?: boolean;
};

/**
 * 日期
 */
export type DataObjectDate = {
  type: DataObjectTypeName.Date;
  defaultValue?: `${number}-${number}-${number}`;
};

/**
 * 日期时间
 */
export type DataObjectDateTime = {
  type: DataObjectTypeName.DateTime;
  defaultValue?: `${number}-${number}-${number} ${number}:${number}:${number}`;
};

/**
 * 时间戳
 */
export type DataObjectTimestamp = {
  type: DataObjectTypeName.Timestamp;
  /**
   * 从 1970-01-01 00:00:00 开始的时间戳
   */
  defaultValue?: `${number}-${number}-${number} ${number}:${number}:${number}`;
};

/**
 * 整数
 */
export type DataObjectInteger = {
  type: DataObjectTypeName.Integer;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * 小数
 */
export type DataObjectDecimal = {
  type: DataObjectTypeName.Decimal;
  defaultValue?: number;

  /**
   * 精度, 默认 10, 最大为 65
   */
  precision?: number;

  /**
   * 小数位, 默认 0
   */
  scale?: number;
};

/**
 * 长整型
 */
export type DataObjectLong = {
  type: DataObjectTypeName.Long;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * Double
 */
export type DataObjectDouble = {
  type: DataObjectTypeName.Double;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * Float
 */
export type DataObjectFloat = {
  type: DataObjectTypeName.Float;
  defaultValue?: number;

  /**
   * 是否自增
   */
  autoIncrement?: boolean;
};

/**
 * String
 */
export type DataObjectString = {
  type: DataObjectTypeName.String;
  defaultValue?: string;

  /**
   * 对应 varchar
   * 0 to 65,535.
   * 默认 255
   */
  length?: number;
};

/**
 * Text
 */
export type DataObjectText = {
  type: DataObjectTypeName.Text;
};

/**
 * 长文本
 */
export type DataObjectLongText = {
  type: DataObjectTypeName.LongText;
};

export enum DataObjectReferenceCardinality {
  OneToOne = 'OneToOne',
  OneToMany = 'OneToMany',
  ManyToOne = 'ManyToOne',
  ManyToMany = 'ManyToMany',
}

/**
 * 字段引用
 */
export type DataObjectReference = {
  type: DataObjectTypeName.Reference;

  /**
   * 目标对象 ID
   */
  target?: string;

  /**
   * 目标对象字段 ID
   */
  targetProperty?: string;

  /**
   * 基数关系
   */
  cardinality?: DataObjectReferenceCardinality;
};

export type DataObjectJSON = {
  type: DataObjectTypeName.JSON;
};

/**
 * 所有支持的数据类型
 */
export type DataObjectTypeDSL =
  | DataObjectBoolean
  | DataObjectDate
  | DataObjectDateTime
  | DataObjectTimestamp
  | DataObjectInteger
  | DataObjectDecimal
  | DataObjectLong
  | DataObjectDouble
  | DataObjectFloat
  | DataObjectString
  | DataObjectText
  | DataObjectLongText
  | DataObjectJSON
  | DataObjectReference;

/**
 * 数据对象字段
 */
export interface DataObjectPropertyDSL<T extends DataObjectTypeDSL = DataObjectTypeDSL> extends NameDSL {
  /**
   * 表字段名，默认为 name 的 kebab_case 模式
   */
  propertyName?: string;

  /**
   * 字段类型
   */
  type: T;

  /**
   * 是否非空
   */
  notNull?: boolean;

  /**
   * 是否为主键
   */
  primaryKey?: boolean;
}

export enum DataObjectIndexType {
  Normal = 'Normal',
  Unique = 'Unique',
  // PrimaryKey = 'Primary',
  FullText = 'FullText',
}

export enum DataObjectIndexMethod {
  BTREE = 'BTREE',
  HASH = 'HASH',
}

/**
 * 索引描述
 */
export interface DataObjectIndexDSL extends NameDSL {
  /**
   * 索引类型
   */
  type: DataObjectIndexType;

  /**
   * 索引字段
   */
  properties: string[];

  /**
   * 索引方法
   */
  method: DataObjectIndexMethod;
}

/**
 * 数据对象
 */
export interface DataObjectDSL extends NameDSL {
  /**
   * 表名称，默认为 name 的 kebab_case 模式
   */
  tableName?: string;

  /**
   * 字段
   */
  properties: DataObjectPropertyDSL[];

  /**
   * 索引
   */
  indexes: DataObjectIndexDSL[];
}
