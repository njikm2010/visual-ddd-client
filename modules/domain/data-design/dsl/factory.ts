import { createNameDSL } from '@/modules/domain/domain-design/dsl/factory';

import {
  DataObjectBoolean,
  DataObjectDate,
  DataObjectDateTime,
  DataObjectDecimal,
  DataObjectDouble,
  DataObjectDSL,
  DataObjectFloat,
  DataObjectIndexDSL,
  DataObjectIndexMethod,
  DataObjectIndexType,
  DataObjectInteger,
  DataObjectLong,
  DataObjectLongText,
  DataObjectPropertyDSL,
  DataObjectReference,
  DataObjectString,
  DataObjectText,
  DataObjectTimestamp,
  DataObjectTypeDSL,
  DataObjectTypeName,
} from './dsl';

export function createDataObjectBoolean(): DataObjectBoolean {
  return {
    type: DataObjectTypeName.Boolean,
    defaultValue: undefined,
  };
}

export function createDataObjectDate(): DataObjectDate {
  return {
    type: DataObjectTypeName.Date,
    defaultValue: undefined,
  };
}

export function createDataObjectDateTime(): DataObjectDateTime {
  return {
    type: DataObjectTypeName.DateTime,
    defaultValue: undefined,
  };
}

export function createDataObjectTimestamp(): DataObjectTimestamp {
  return {
    type: DataObjectTypeName.Timestamp,
    defaultValue: undefined,
  };
}

export function createDataObjectInteger(): DataObjectInteger {
  return {
    type: DataObjectTypeName.Integer,
    defaultValue: undefined,
    autoIncrement: false,
  };
}

export function createDataObjectDecimal(): DataObjectDecimal {
  return {
    type: DataObjectTypeName.Decimal,
    defaultValue: undefined,
    precision: undefined,
    scale: undefined,
  };
}

export function createDataObjectLong(): DataObjectLong {
  return {
    type: DataObjectTypeName.Long,
    defaultValue: undefined,
    autoIncrement: false,
  };
}

export function createDataObjectDouble(): DataObjectDouble {
  return {
    type: DataObjectTypeName.Double,
    defaultValue: undefined,
    autoIncrement: false,
  };
}

export function createDataObjectFloat(): DataObjectFloat {
  return {
    type: DataObjectTypeName.Float,
    defaultValue: undefined,
    autoIncrement: false,
  };
}

export function createDataObjectString(): DataObjectString {
  return {
    type: DataObjectTypeName.String,
    defaultValue: undefined,
    length: undefined,
  };
}

export function createDataObjectText(): DataObjectText {
  return {
    type: DataObjectTypeName.Text,
    defaultValue: undefined,
  };
}

export function createDataObjectLongText(): DataObjectLongText {
  return {
    type: DataObjectTypeName.LongText,
    defaultValue: undefined,
  };
}

export function createDataObjectReference(): DataObjectReference {
  return {
    type: DataObjectTypeName.Reference,
    target: undefined,
    targetProperty: undefined,
    cardinality: undefined,
  };
}

export function createDataObjectType(type: DataObjectTypeName): DataObjectTypeDSL {
  switch (type) {
    case DataObjectTypeName.Boolean:
      return createDataObjectBoolean();
    case DataObjectTypeName.Date:
      return createDataObjectDate();
    case DataObjectTypeName.DateTime:
      return createDataObjectDateTime();
    case DataObjectTypeName.Timestamp:
      return createDataObjectTimestamp();
    case DataObjectTypeName.Integer:
      return createDataObjectInteger();
    case DataObjectTypeName.Decimal:
      return createDataObjectDecimal();
    case DataObjectTypeName.Long:
      return createDataObjectLong();
    case DataObjectTypeName.Double:
      return createDataObjectDouble();
    case DataObjectTypeName.Float:
      return createDataObjectFloat();
    case DataObjectTypeName.String:
      return createDataObjectString();
    case DataObjectTypeName.Text:
      return createDataObjectText();
    case DataObjectTypeName.LongText:
      return createDataObjectLongText();
    case DataObjectTypeName.Reference:
      return createDataObjectReference();
  }
}

/**
 * 创建属性
 * @param type
 * @returns
 */
export function createDataObjectPropertyDSL(
  type: DataObjectTypeName = DataObjectTypeName.String
): DataObjectPropertyDSL {
  return {
    ...createNameDSL({ wordCase: 'camelCase', title: false }),
    propertyName: undefined,
    notNull: false,
    type: createDataObjectType(type),
  };
}

/**
 * 创建索引
 * @returns
 */
export function createDataObjectIndexDSL(): DataObjectIndexDSL {
  return {
    ...createNameDSL({ wordCase: 'camelCase', title: true }),
    type: DataObjectIndexType.Normal,
    properties: [],
    method: DataObjectIndexMethod.BTREE,
  };
}

/**
 * 创建数据对象
 * @returns
 */
export function createDataObjectDSL(): DataObjectDSL {
  const base = createNameDSL({ wordCase: 'CamelCase', title: true });

  return {
    ...base,
    tableName: undefined,
    properties: [createDataObjectPropertyDSL()],
    indexes: [],
  };
}
