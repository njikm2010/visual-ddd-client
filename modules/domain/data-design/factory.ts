import { BaseEditorModelOptions, BuiltinShapeName } from '@/lib/editor';

import { DataObjectName } from './dsl';
import { DataObjectEditorModel } from './model';

/**
 * 模型构造
 * @param options
 * @returns
 */
export function createDataObjectEditorModel(
  options: Omit<BaseEditorModelOptions, 'shapeList' | 'whitelist' | 'scopeId'>
) {
  const shapeList = [DataObjectName.DataObject, BuiltinShapeName.Comment];

  const whitelist = shapeList;

  return new DataObjectEditorModel({ ...options, scopeId: 'data-object', shapeList, whitelist, activeScope: false });
}
