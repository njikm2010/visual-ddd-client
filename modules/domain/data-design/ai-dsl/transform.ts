import { DataObjectDSL, DataObjectTypeDSL, DataObjectTypeName, DataObjectPropertyDSL } from '../dsl/dsl';
import { NameDSL } from '../../domain-design/dsl/dsl';
import { UntitledInHumanReadable } from '../../domain-design/dsl/constants';
import snakeCase from 'lodash/snakeCase';
import { assert } from '@/lib/utils';

function getTableName(table: DataObjectDSL) {
  return table.tableName || snakeCase(table.name);
}

function getPropertyName(property: DataObjectPropertyDSL) {
  return property.propertyName || snakeCase(property.name);
}

function stringifyType(type: DataObjectTypeDSL, tables: DataObjectDSL[]) {
  if (type.type !== DataObjectTypeName.Reference) {
    return type.type;
  }

  const refTable = tables.find(i => i.uuid === type.target);

  assert(refTable, `未找到引用的数据表 ${type.target}`);

  const refProperty = refTable.properties.find(i => i.uuid === type.targetProperty);

  assert(refProperty, `未找到引用的数据表 ${refTable.name} 的属性 ${type.targetProperty}`);

  return `${type.type}(${getTableName(refTable)}.${getPropertyName(refProperty)})`;
}

function getComment(data: NameDSL) {
  if (data.title && !data.title.includes(UntitledInHumanReadable)) {
    return data.title;
  }
  return undefined;
}

function stringifyProperty(property: DataObjectPropertyDSL, table: DataObjectDSL, tables: DataObjectDSL[]): string {
  const type = stringifyType(property.type, tables);
  const modifier: string[] = [];

  if (property.primaryKey) {
    modifier.push('PrimaryKey');
  } else if (property.notNull) {
    modifier.push('NotNull');
  }

  // TODO: 支持默认值
  const comment = getComment(property);

  return `  ${getPropertyName(property)}: ${type}${modifier.length ? ', ' : ''}${modifier.join(', ')};${
    comment ? ` #${comment}` : ''
  }`;
}

function stringifyTable(table: DataObjectDSL, tables: DataObjectDSL[]): string {
  const comment = getComment(table);
  return `${comment ? `#${comment}\n` : ''}Table ${getTableName(table)} (\n${table.properties
    .map(i => {
      return stringifyProperty(i, table, tables);
    })
    .join('\n')}
)`;
}

function stringifyIndex(table: DataObjectDSL): string {
  if (table.indexes.length) {
    let dsl = '';
    for (const index of table.indexes) {
      if (!index.properties.length) {
        continue;
      }

      const idx = `Index ${getTableName(table)}.${index.name} (
	type: ${index.type};
	method: ${index.method};
	fields: ${index.properties
    .map(p => {
      const property = table.properties.find(i => i.uuid === p);
      assert(property, `未找到属性 ${p}`);

      return getPropertyName(property);
    })
    .join(', ')};
)`;
      dsl += idx + '\n\n';
    }

    return dsl;
  }

  return '';
}

/**
 * 将数据表 DSL 转换为 AI 容易识别的，简洁的格式
 */
export function transform(tables: DataObjectDSL[], allTables: DataObjectDSL[]): string {
  let dsl = '';

  for (const table of tables) {
    const tbl = stringifyTable(table, allTables);
    const index = stringifyIndex(table);

    dsl += `${tbl}\n\n${index}`;
  }

  return dsl;
}
