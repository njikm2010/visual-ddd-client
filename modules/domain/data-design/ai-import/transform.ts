import { AIImport } from './dsl';
import { DataObjectDSL, DataObjectReference, DataObjectTypeName } from '../dsl/dsl';
import { createDataObjectDSL, createDataObjectPropertyDSL, createDataObjectType } from '../dsl/factory';
import { assert } from '@/lib/utils';

function isReference(
  type: AIImport.DataObjectNormalType | AIImport.DataObjectReferenceType
): type is AIImport.DataObjectReferenceType {
  return type.type === DataObjectTypeName.Reference;
}

export function transform(data: AIImport.DataObject[]): DataObjectDSL[] {
  const pending: Function[] = [];
  const list: DataObjectDSL[] = [];

  for (const item of data) {
    const object = createDataObjectDSL();
    object.title = item.title;
    object.name = item.name;
    object.properties = [];

    for (const prop of item.properties) {
      const property = createDataObjectPropertyDSL();
      property.name = prop.name;
      property.title = prop.title;

      if (prop.primaryKey) {
        property.notNull = true;
        property.primaryKey = true;
      }

      property.type = createDataObjectType(prop.type.type ?? DataObjectTypeName.String);

      if (isReference(prop.type)) {
        const type = prop.type;
        pending.push(() => {
          const obj = list.find(i => i.name === type.target);
          assert(obj, `引用类型解析错误, 未找到对象 ${type.target}`);

          const field = obj.properties.find(i => i.name === type.property);

          assert(field, `引用类型解析错误, 未找到字段 ${type.target}.${type.property}`);

          const ref = property.type as DataObjectReference;
          ref.target = obj.uuid;
          ref.targetProperty = field.uuid;
          ref.cardinality = type.cardinality;
        });
      }

      object.properties.push(property);
    }

    list.push(object);
  }

  for (const fn of pending) {
    fn();
  }

  return list;
}
