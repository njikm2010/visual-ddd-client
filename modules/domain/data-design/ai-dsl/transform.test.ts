import { createDataObjectDSL, createDataObjectIndexDSL, createDataObjectPropertyDSL } from '../dsl/factory';
import { DataObjectPropertyDSL, DataObjectReference, DataObjectString, DataObjectTypeName } from '../dsl/dsl';
import { transform } from './transform';

describe('transform', () => {
  const inString = createDataObjectPropertyDSL(DataObjectTypeName.String);
  inString.name = 'inString';
  inString.title = '字符串';
  (inString.type as DataObjectString).defaultValue = 'hello';

  const inInteger = createDataObjectPropertyDSL(DataObjectTypeName.Integer);
  inInteger.name = 'inInteger';
  inInteger.title = '整数';

  const inPrimaryKey = createDataObjectPropertyDSL(DataObjectTypeName.Long);
  inPrimaryKey.name = 'inPrimaryKey';
  inPrimaryKey.title = '主键';
  inPrimaryKey.primaryKey = true;

  const inNotNull = createDataObjectPropertyDSL(DataObjectTypeName.Date);
  inNotNull.name = 'inNotNull';
  inNotNull.title = '非空';
  inNotNull.notNull = true;

  const customPropertyName = createDataObjectPropertyDSL(DataObjectTypeName.String);
  customPropertyName.name = 'customPropertyName';
  customPropertyName.title = '自定义属性名';
  customPropertyName.propertyName = 'my_name';

  test('simpleTable', () => {
    const t = createDataObjectDSL();
    t.name = 'MyTable';
    t.title = '我的表格';
    t.properties = [];
    t.properties.push(inPrimaryKey, inString, inInteger, inNotNull, customPropertyName);

    expect(transform([t], [t])).toMatchSnapshot();

    // for conception
    expect(transform([t], [t], { useCase: 'conception' })).toMatchSnapshot();
  });

  test('index', () => {
    const t = createDataObjectDSL();
    t.name = 'MyTable';
    t.properties = [];
    t.properties.push(inPrimaryKey, inString, inInteger, inNotNull, customPropertyName);

    const index = createDataObjectIndexDSL();
    index.name = 'MyIndex';
    index.properties = [inPrimaryKey.uuid, inString.uuid];

    const emptyIndex = createDataObjectIndexDSL();
    emptyIndex.name = 'EmptyIndex';

    t.indexes = [index, emptyIndex];

    expect(transform([t], [t])).toMatchSnapshot();
  });

  test('tableReference', () => {
    const a = createDataObjectDSL();
    a.name = 'A';
    a.properties = [inPrimaryKey, inString];

    const b = createDataObjectDSL();
    const ref = createDataObjectPropertyDSL(DataObjectTypeName.Reference) as DataObjectPropertyDSL<DataObjectReference>;
    ref.type.target = a.uuid;
    ref.type.targetProperty = inPrimaryKey.uuid;

    b.properties = [inPrimaryKey, ref];

    expect(transform([a, b], [a, b])).toMatchSnapshot();
  });
});
