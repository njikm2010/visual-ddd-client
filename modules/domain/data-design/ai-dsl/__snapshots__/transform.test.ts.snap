// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform index 1`] = `
"Table my_table (
  in_primary_key: Long, PrimaryKey; #主键
  in_string: String; #字符串
  in_integer: Integer; #整数
  in_not_null: Date, NotNull; #非空
  my_name: String; #自定义属性名
)

Index my_table.MyIndex (
	type: Normal;
	method: BTREE;
	fields: in_primary_key, in_string;
)

"
`;

exports[`transform simpleTable 1`] = `
"#我的表格
Table my_table (
  in_primary_key: Long, PrimaryKey; #主键
  in_string: String; #字符串
  in_integer: Integer; #整数
  in_not_null: Date, NotNull; #非空
  my_name: String; #自定义属性名
)

"
`;

exports[`transform tableReference 1`] = `
"Table a (
  in_primary_key: Long, PrimaryKey; #主键
  in_string: String; #字符串
)

Table untitled (
  in_primary_key: Long, PrimaryKey; #主键
  untitled: Reference(a.in_primary_key);
)

"
`;
